File: packages/telemetry/src/__tests__/run-preflight.test.ts
import { describe, expect, it, vi } from 'vitest';
import { runPreflightChecks } from '../preflight/run-preflight.js';

const runContext = {
	flagKey: 'demo-feature-flag',
	treatmentVariationId: 'treatment-var',
	controlVariationId: 'control-var',
};

const policies = [
	{
		metricType: 'error_rate',
		threshold: 0.01,
		serviceName: 'demo-service',
		minSampleSize: 100,
	},
];

function vector(value: string) {
	return { resultType: 'vector', result: [{ metric: {}, value: [0, value] }] };
}

describe('runPreflightChecks', () => {
	it('passes when treatment/control samples and user context are present', async () => {
		const client = {
			config: { baseUrl: 'http://localhost:9090' },
			queryInstant: vi
				.fn()
				.mockResolvedValueOnce(vector('200'))
				.mockResolvedValueOnce(vector('200')),
		};

		const report = await runPreflightChecks(client, policies, runContext);
		expect(report.status).toBe('pass');
		expect(report.checks.map((check) => check.id)).toEqual([
			'metric_flow_treatment',
			'metric_flow_control',
			'min_sample_treatment',
			'min_sample_control',
			'context_kind_user',
		]);
	});

	it('fails when treatment metric flow is missing', async () => {
		const client = {
			config: { baseUrl: 'http://localhost:9090' },
			queryInstant: vi
				.fn()
				.mockResolvedValueOnce({ resultType: 'vector', result: [] })
				.mockResolvedValueOnce(vector('200')),
		};

		const report = await runPreflightChecks(client, policies, runContext);
		expect(report.status).toBe('fail');
		expect(report.blockReason).toBeTruthy();
		expect(
			report.checks.find((check) => check.id === 'metric_flow_treatment')?.status,
		).toBe('fail');
	});
});

File: kotlin/modules/telemetry/src/test/kotlin/com/ffpromo/telemetry/preflight/RunPreflightTest.kt
package com.ffpromo.telemetry.preflight

import com.ffpromo.contracts.GatePolicyInput
import com.ffpromo.contracts.GateRunContext
import com.ffpromo.contracts.MetricType
import com.ffpromo.contracts.PreflightCheckStatus
import com.ffpromo.telemetry.client.PrometheusInstantQueryData
import com.ffpromo.telemetry.support.QueuePrometheusClient
import com.ffpromo.telemetry.support.vector
import kotlinx.coroutines.test.runTest
import kotlinx.serialization.json.buildJsonArray
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test

class RunPreflightTest {
    private val runContext = GateRunContext(
        flagKey = "demo-feature-flag",
        treatmentVariationId = "treatment-var",
        controlVariationId = "control-var",
    )

    private val policies = listOf(
        GatePolicyInput(
            metricType = MetricType.error_rate,
            threshold = 0.01,
            serviceName = "demo-service",
            minSampleSize = 100,
        ),
    )

    @Test
    fun `passes when treatment and control samples and user context are present`() = runTest {
        val client = QueuePrometheusClient(mutableListOf(vector("200"), vector("200")))
        val report = runPreflightChecks(client, policies, runContext)
        assertEquals(PreflightCheckStatus.pass, report.status)
        assertEquals(
            listOf(
                "metric_flow_treatment",
                "metric_flow_control",
                "min_sample_treatment",
                "min_sample_control",
                "context_kind_user",
            ),
            report.checks.map { it.id },
        )
    }

    @Test
    fun `fails when treatment metric flow is missing`() = runTest {
        val client = QueuePrometheusClient(
            mutableListOf(
                PrometheusInstantQueryData(resultType = "vector", result = buildJsonArray {}),
                vector("200"),
            ),
        )
        val report = runPreflightChecks(client, policies, runContext)
        assertEquals(PreflightCheckStatus.fail, report.status)
        assertTrue(!report.blockReason.isNullOrBlank())
        assertEquals(
            PreflightCheckStatus.fail,
            report.checks.find { it.id == "metric_flow_treatment" }?.status,
        )
    }
}
